scalalib/src/mill/scalalib/TestModuleUtil.scala

```diff
@@ -173,15 +209,217 @@ private[scalalib] object TestModuleUtil {
+        }
+    }
+
+    val jobs = Task.ctx() match {
+      case j: Ctx.Jobs => j.jobs
+      case _ => 1
+    }
+
+    val maxProcessLength = jobs.toString.length
+    val groupLength = groupFolderData.length
+    val maxGroupLength = groupLength.toString.length
+
+    val subprocessFutures = for {
+      processIndex <- 0 until jobs
+      groupIndex <- 0 until groupLength
+      (baseFolder, selectorFolder, numTests, groupPromptMessage) = groupFolderData(groupIndex)
+      if (processIndex == 0) || (processIndex < numTests)
+    } yield {
+      val paddedGroupIndex = mill.internal.Util.leftPad(groupIndex.toString, maxGroupLength, '0')
+      val paddedProcessIndex =
+        mill.internal.Util.leftPad(processIndex.toString, maxProcessLength, '0')
+      val processFolder = baseFolder / processIndex.toString
+
+      Task.fork.async(
+        processFolder,
+        s"${paddedGroupIndex}-${paddedProcessIndex}",
+        groupPromptMessage
+      ) { logger =>
+        // force run when processIndex == 0 (first subprocess), even if there are no tests to run
+        // to force the process to go through the test framework setup/teardown logic
+        s"${baseFolder.last}" -> runTestRunnerSubprocess(
+          groupIndex,
+          processIndex,
+          processFolder,
+          selectorFolder,
+          force = processIndex == 0,
+          logger
+        )
+      }
+    }
+
+    val logScheduler = new java.util.concurrent.ScheduledThreadPoolExecutor(1)
+
+    logScheduler.scheduleWithFixedDelay(
+      new Runnable {
+        override def run(): Unit = {
+          stealLogMap.forEach { case (_, (stealLog, stealLogCallback)) =>
+            try {
+              os.read.lines(stealLog).lastOption.foreach(stealLogCallback)
+            } finally ()
+          }
         }
+      },
+      0,
+      1,
+      java.util.concurrent.TimeUnit.SECONDS
+    )
+
+    val outputs = Task.fork.awaitAll(subprocessFutures)
+
+    logScheduler.shutdown()
+
+    val subprocessResult = {
+      val failMap = mutable.Map.empty[String, String]
+      val successMap = mutable.Map.empty[String, (String, Seq[TestResult])]
+
+      outputs.foreach {
+        case (name, Some(Result.Failure(v))) => failMap.updateWith(name) {
+            case Some(old) => Some(old + " " + v)
+            case None => Some(v)
+          }
+        case (name, Some(Result.Success((msg, results)))) => successMap.updateWith(name) {
+            case Some((oldMsg, oldResults)) => Some((oldMsg + " " + msg, oldResults ++ results))
+            case None => Some((msg, results))
+          }
+        case _ => ()
+      }
+
+      if (failMap.nonEmpty) {
+        Result.Failure(failMap.values.mkString("\n"))
+      } else {
+        Result.Success((
+          successMap.values.map(_._1).mkString("\n"),
+          successMap.values.flatMap(_._2).toSeq
+        ))
+      }
     }
+
+    subprocessResult
+  }
+
+}
+
+private[scalalib] object TestModuleUtil {
+
+  def runTests(
```

Comment:
No need this forwarder, let's just have the callsites use `new TestModuleUtil` directly